Docs: Add Getting started with Grafana Mimir using the Helm chart (#2175)
grafana/mimir: docs/sources/operators-guide/getting-started-helm-charts/_index.md

  wal_directory: /tmp/grafana-agent/wal

  configs:
    - name: agent
      scrape_configs:
        - job_name: agent
          static_configs:
            - targets: ["127.0.0.1:12345"]
      remote_write:
        - url: http://<ingress-host>/api/v1/push
```

START HERE:

## Query data in Grafana

You can either [deploy Grafana Mimir on Kubernetes](https://grafana.com/docs/grafana/latest/setup-grafana/installation/kubernetes/)
or get a test instance of a local Grafana server up and running
quickly by using Docker:

```bash
docker run --rm --name=grafana --network=host grafana/grafana
```

### Add Grafana Mimir as a Prometheus data source

1. In a browser, go to the Grafana server at [http://localhost:3000/datasources](http://localhost:3000/datasources).
1. Sign in using the default username `admin` and password `admin`.
1. Configure a new Prometheus data source to query the local Grafana Mimir server using the following settings:

   | Field | Value                                                                |
   | ----- | -------------------------------------------------------------------- |
   | Name  | Mimir                                                                |

>    | URL   | [http://localhost:9009/prometheus](http://localhost:9009/prometheus) |

Reviewer: ```suggestion
   | URL   | [http://<ingress-host>/prometheus](http://localhost:9009/prometheus) |
```

---
title: "Getting started with Grafana Mimir using the Helm chart"
menuTitle: "Getting started using the Helm chart"
description: "Learn how to get started with Grafana Mimir using the Helm chart."
weight: 20
---

# Getting started with Grafana Mimir using the Helm chart

<!-- What is this about? TODO -->

## Before you begin

The instructions that follow are common across any flavor of Kubernetes.

The information that follows assumes that you are able to install Kubernetes, and configure and operate it.

Install the following software:

- Either a [Prometheus server](https://prometheus.io/docs/prometheus/latest/installation/) or [Grafana Agent](https://grafana.com/docs/grafana-cloud/agent/#installing-the-grafana-agent).

> <!-- TBD, Krajo to figure out: - Verify that you have enough memery overall and number of cores. Krajo needs a couple of days for this answer to come in. -->

Reviewer: A minimal choice is to say that you have 2 cores (2000m) and 8GiB memory available, which means a laptop with something like 4 cores and 16 GiB RAM is enough to get started, but this is not recommended for production as in production we start from 3 nodes(!) for redundancy.

   <release-name>-mimir-ruler-745fff6c6d-8swsv                1/1     Running     0          148m
   <release-name>-mimir-alertmanager-0                        1/1     Running     0          148m
   <release-name>-mimir-compactor-0                           1/1     Running     0          148m
   <release-name>-mimir-ingester-1                            1/1     Running     0          147m
   <release-name>-mimir-ingester-2                            1/1     Running     0          147m
   <release-name>-mimir-store-gateway-0                       1/1     Running     0          146m
   <release-name>-mimir-ingester-0                            1/1     Running     0          144m
   ```

1. What until all of the pods have a status of `Running` or `Completed`, which might take a few minutes.

## Configure Prometheus to write to Grafana Mimir

Add the following YAML snippet to your Prometheus configuration file and restart the Prometheus server:

```yaml
remote_write:
  - url: http://<ingress-host>/api/v1/push
```

The configuration for a Prometheus server that scrapes itself and writes those metrics to Grafana Mimir looks similar to this:

```yaml
remote_write:

>   - url: http://<ingress-host>/api/v1/push

Reviewer 2: Using ingress-host will only work when Prometheus runs outside of Kubernetes server. Do we assume such setup?

Reviewer: We're trying to cover both cases, but maybe it's too ambitious . There's a note further up that says what URL to use if you don't have ingress.

Reviewer: As discussed in real life, we'll have two chapters for (outside the cluster with ingress) and (inside the cluster using metamonitoring to loop back on itself) CC: @osg-grafana  @dimitarvdimitrov 

Should be a pretty cool and self contained way. We can skip Prometheus , or just mention that the same kind of remote write setup works for it.

scrape_configs:
  - job_name: prometheus
    honor_labels: true
    static_configs:

>       - targets: ["localhost:9090"]

Reviewer 2: How does "localhost:9090" map to all Mimir pods? Will it only scrape single target?

When running in Kubernetes, shouldn't we use features that allow Prometheus to find all pods and scrape them all? 

(In GrafanaCon 2022 demo we used Grafana Agent Operator, created instance of Grafana Agent *inside* k8s cluster, and let it scrape all Mimir pods. You can see the demo at https://grafana.com/go/grafanaconline/2022/grafana-mimir-migrate-your-metrics-in-minutes/, starting ~9m10s.)

Reviewer: I've tested Grafana Agent running outside the cluster and sending logs to Mimir inside the cluster. Haven't tried running Prometheus yet, but you're right it won't find Mimir on localhost 9090

Reviewer: I've just tested this and prometheus does scrape itself on port 9090, so it makes sense to have an example with Prometheus as well. Got a one liner docker command working.

```

## Configure Grafana Agent to write to Grafana Mimir

> 

Reviewer: We should add some reference to docs about installing Grafana Agent or give a simple docker example.

---
title: "Getting started with Grafana Mimir using the Helm chart"
menuTitle: "Getting started using the Helm chart"
description: "Learn how to get started with Grafana Mimir using the Helm chart."
weight: 20
---

# Getting started with Grafana Mimir using the Helm chart

<!-- What is this about? TODO -->
<!-- TODO introduce the notation of <...> to mean that the user needs to set their own value here. -->

## Before you begin

The instructions that follow are common across any flavor of Kubernetes.

The information that follows assumes that you are able to install Kubernetes, and configure and operate it.

Hardware requirements:

- For this tutorial a single Kubernetes node with a minimum of 4 cores and 16GiB RAM should be sufficient. Note that production setups start from 3 nodes for redundancy!

Install the following software:

> - Either a [Prometheus server](https://prometheus.io/docs/prometheus/latest/installation/) or [Grafana Agent](https://grafana.com/docs/grafana-cloud/agent/#installing-the-grafana-agent).

Reviewer: Hmm, we should phrase this in an other way, cause we'll be giving two different ways of doing things later, so this is pretty under-defined here. Remove from here?

- Helm 3 or higher
- Kubernetes 1.10 or higher
- The kubectl command for your version of Kubernetes, and an understanding of what the command does.

Verify that you have:

- Access to the Kubernetes cluster
- Persistent storage is enabled in the Kubernetes cluster and there is a default storage class set up
- DNS service works in the Kubernetes cluster
- An ingress controller is set up in the Kubernetes cluster

  **Note:** Although this is not strictly necessary, if you want to access Mimir from outside of the Kubernetes cluster, you will need an ingress.

## Install the Helm chart in a custom namespace

Using a custom namespace solves problems later on because you do not have to overwrite the default namespace.

1. [Create a unique Kubernetes namespace](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#creating-a-new-namespace):

   ```console
   kubectl create namespace <namespace>
   ```

1. Set up a Helm repository using the following commands:

   ```console
   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

>    > **Note:** The Helm chart at [https://grafana.github.io/helm-charts](https://grafana.github.io/helm-charts) is a publication of the source code at [**grafana/mimir**](https://github.com/grafana/mimir/tree/main/operations/helm/charts/mimir-distributed).

Reviewer: We could also mention here that bug reports, feature requests and contributions should go into the mimir repo.

#### Add Grafana Mimir as a Prometheus data source

1. In a browser, go to the Grafana server at [http://localhost:3000/datasources](http://localhost:3000/datasources).
1. Sign in using the default username `admin` and password `admin`.
1. Configure a new Prometheus data source to query the local Grafana Mimir server using the following settings:

   | Field | Value                              |
   | ----- | ---------------------------------- |
   | Name  | Mimir                              |
   | URL   | http://\<ingress-host\>/prometheus |

To add a data source, refer to [Add a data source](https://grafana.com/docs/grafana/latest/datasources/add-a-data-source/).

### Verify success

When you have completed the tasks in this getting started guide, you can query metrics in [Grafana Explore](https://grafana.com/docs/grafana/latest/explore/)
as well as create dashboard panels using the newly configured Grafana Mimir data source.

## Accessing Grafana Mimir from inside the Kubernetes cluster

This chapter does not assume that an ingress is set up, but of course it does not prohibit it either. You can mix the two approaches as needed.

### Enable Meta Monitoring in Grafana Mimir

> Grafana Mimir meta monitoring collects metrics and or logs about Grafana Mimir itself. It's primary purpose is to send meta monitoring information to some external receiver, for example a [free tier Grafana Metrics account](https://grafana.com/metrics/). In this example we'll use it to collect and send metrics into itself for testing.

Reviewer: @dimitarvdimitrov  I've written some words here about meta monitoring, but I'm pretty sure you have something better for a short intro on meta monitoring :)

1. Set up a Helm repository using the following commands:

   ```console
   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

   > **Note:** The Helm chart at [https://grafana.github.io/helm-charts](https://grafana.github.io/helm-charts) is a publication of the source code at [**grafana/mimir**](https://github.com/grafana/mimir/tree/main/operations/helm/charts/mimir-distributed).

1. Configure an ingress by creating a YAML file of Helm values, such as `custom.yaml` and adding the following configuration:

   ```yaml
   nginx:
     ingress:
       enabled: true
       ingressClassName: nginx
       hosts:
         - host: <ingress-host>
           paths:
             - path: /
               pathType: Prefix
       tls:
         # empty, disabled.
   ```

   An ingress enables you to externally access a Kubernetes cluster.
   Replace _`<ingress-host>`_ with a suitable hostname that DNS can resolve to the external IP address of the Kubernetes cluster. For more information, see [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/).

>    > **Note:** Without using an ingress, it is still possible to access Grafana Mimir from inside the cluster.

Author: - [ ] Add a link to a resource where the user can get more information.

---
title: "Getting started with Grafana Mimir using the Helm chart"
menuTitle: "Getting started using the Helm chart"
description: "Learn how to get started with Grafana Mimir using the Helm chart."
weight: 20
---

# Getting started with Grafana Mimir using the Helm chart

> The Helm chart allows you to configure, install, and upgrade Grafana Mimir within a Kubernetes cluster.

Reviewer: ```suggestion
The [Helm](https://helm.sh/) chart allows you to configure, install, and upgrade Grafana Mimir within a Kubernetes cluster.
```

## Before you begin

The instructions that follow are common across any flavor of Kubernetes. They also assume that you know how to install a Kubernetes cluster, and configure and operate it.

It also assumes that you have an understanding of what the `kubectl` command does.

> **Caution:** Do not use this getting-started procedure in a production environment.

Hardware requirements:

- A single Kubernetes node with a minimum of 4 cores and 16GiB RAM

Software requirements:

> - Kubernetes 1.10 or higher

Reviewer: ```suggestion
- Kubernetes 1.20 or higher
```

Merged #2297

- The `kubectl` command for your version of Kubernetes
- Helm 3 or higher

Verify that you have:

- Access to the Kubernetes cluster

> - Persistent storage is enabled in the Kubernetes cluster, which has a default storage class set up

Reviewer: ```suggestion
- Persistent storage is enabled in the Kubernetes cluster, which has a default storage class set up. See [Change the default StorageClass](https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/) in the Kubernetes documentation.
```

- DNS service works in the Kubernetes cluster

> - An ingress controller is set up in the Kubernetes cluster

Reviewer: ```suggestion
- An ingress controller is set up in the Kubernetes cluster, for example [ingress-nginx](https://kubernetes.github.io/ingress-nginx/)
```

   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

   > **Note:** The Helm chart at [https://grafana.github.io/helm-charts](https://grafana.github.io/helm-charts) is a publication of the source code at [**grafana/mimir**](https://github.com/grafana/mimir/tree/main/operations/helm/charts/mimir-distributed).

1. Configure an ingress:

   a. Create a YAML file of Helm values called `custom.yaml`.

   b. Add the following configuration to the file:

   ```yaml
   nginx:
     ingress:
       enabled: true
       ingressClassName: nginx
       hosts:
         - host: <ingress-host>
           paths:
             - path: /
               pathType: Prefix
       tls:
         # empty, disabled.
   ```

   An ingress enables you to externally access a Kubernetes cluster.
   Replace _`<ingress-host>`_ with a suitable hostname that DNS can resolve to the external IP address of the Kubernetes cluster. For more information, see [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/).

1. Install Grafana Mimir using the Helm chart:

   ```bash
   helm -n mimir-test install mimir grafana/mimir-distributed -f custom.yaml
   ```

>    > **Note:**: the output of the command contains the write and read URLs necessary for the following steps.

Reviewer 3: ```suggestion
   > **Note:** the output of the command contains the write and read URLs necessary for the following steps.
```

   ```

1. Wait until all of the pods have a status of `Running` or `Completed`, which might take a few minutes.

## Configure Prometheus to write to Grafana Mimir

Make a choice based on whether or not you already have a Prometheus server set up:

- For an existing Prometheus server:

  1. Add the following YAML snippet to your Prometheus configuration file:

     ```yaml
     remote_write:
       - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Prometheus server writes metrics to Grafana Mimir, based on what is defined in the existing `scrape_configs` configuration.

  1. Restart the Prometheus server.

- For a Prometheus server that does not exist yet:

  1. Write the following configuration to a `prometheus.yml` file:

     ```yaml
     remote_write:
       - url: http://<ingress-host>/api/v1/push

     scrape_configs:
       - job_name: prometheus
         honor_labels: true
         static_configs:
           - targets: ["localhost:9090"]
     ```

     In this case, your Prometheus server writes metrics to Grafana Mimir that it scrapes from itself.

  1. Start a Prometheus server by using Docker:

     ```bash

>      docker run --network=host -p 9090:9090  -v <path-to>/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus

Author: ```suggestion
     docker run --network=host -p 9090:9090  -v <absolute-path-to>/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
```

>      docker run --network=host -p 9090:9090  -v <absolute-path-to>/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus

Reviewer 4: Minor thing. Since we are using `--network=host` we can remove `-p 9090:9090` because it will be ignored anyway.

- For an existing Grafana Agent:

  1. Add the following YAML snippet to your Grafana Agent metrics configurations (`metrics.configs`):

     ```yaml
     remote_write:
       - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Grafana Agent will write metrics to Grafana Mimir, based on what is defined in the existing `metrics.configs.scrape_configs` configuration.

  1. Restart the Grafana Agent.

- For a Grafana Agent that does not exist yet:

  1. Write the following configuration to an `agent.yaml` file:

     ```yaml
     server:
       http_listen_port: 12345
       grpc_listen_port: 54321

     metrics:
       wal_directory: /tmp/grafana-agent/wal

       configs:
         - name: agent
           scrape_configs:
             - job_name: agent
               static_configs:
                 - targets: ["127.0.0.1:12345"]
           remote_write:
             - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Grafana Agent writes metrics to Grafana Mimir that it scrapes from itself.

  1. Create an empty directory for the write ahead log (WAL) of the Grafana Agent

  1. Start a Grafana Agent by using Docker:

     ```bash

>      docker run --network=host  -v <path-to-wal-directory>:/etc/agent/data -v <path-to>/agent.yaml:/etc/agent/agent.yaml grafana/agent

Author: ```suggestion
     docker run --network=host  -v <absolute-path-to-wal-directory>:/etc/agent/data -v <path-to>/agent.yaml:/etc/agent/agent.yaml grafana/agent
```

Author: Are both of these paths absolute?

     ```

## Query metrics in Grafana

First install Grafana, and then add Mimir as a Prometheus data source.

1. Start Grafana by using Docker:

   ```bash

>    docker run --rm --name=grafana --network=host grafana/grafana

Reviewer 3: This only works for docker on linux. The below should work on all dockers

```suggestion
   docker run --rm --name=grafana -p 3000:3000 grafana/grafana
```

Reviewer: That was required in my linux setup to get it working with having the ingress hostname in my `/etc/hosts` file.
The other option which might be more portable is using the `--add-host=<ingress-host>:<cluster-external-ip>` option.

Reviewer: Actually add-host doesn't work on Windows either: https://github.com/docker/for-win/issues/1455

Reviewer: The most portable solution is to require a DNS resolve-able `ingress-host` fqdn.  But on linux it's just so much easier to put it into /etc/hosts

Reviewer: For now I propose to primarily document the DNS resolvable way but add notes about /etc/hosts and --network-host for linux

Reviewer 4: If the ingress is in local for testing, we can document to use localdev.me which always resolves to 127.0.0.1.

Reviewer: Just tested without /etc/hosts:
```
docker run -p 9090:9090 --add-host=mimir.krajo:10.64.0.1 -v `pwd`/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
docker run --rm --name=grafana -p 3000:3000 --add-host=mimir.krajo:10.64.0.1 grafana/grafana
```

So no need for /etc/hosts, `--add-host` works fine.

Reviewer: updated without `--network=host` on the mainline

   | URL   | http://\<ingress-host\>/prometheus |

   To add a data source, see [Add a data source](https://grafana.com/docs/grafana/latest/datasources/add-a-data-source/).

1. Verify success:

   You should be able to query metrics in [Grafana Explore](http://localhost:3000/explore),
   as well as create dashboard panels by using your newly configured `Mimir` data source.

## Set up metamonitoring

Grafana Mimir metamonitoring collects metrics or logs, or both,
about Grafana Mimir itself.
In the example that follows, metamonitoring scrapes metrics about
Grafana Mimir itself, and then writes those metrics to the same Grafana Mimir instance.

1. To enable metamonitoring in Grafana Mimir, add the following YAML snippet to your Grafana Mimir `custom.yaml` file:

   ```yaml
   metaMonitoring:
     serviceMonitor:
       enabled: true
     grafanaAgent:
       enabled: true
       installOperator: true
       metrics:
         additionalRemoteWriteConfigs:
           - url: "http://mimir-nginx.mimir-test.svc:80/api/v1/push"
   ```

1. Upgrade Grafana Mimir by using the `helm` command:

   ```bash
   helm -n mimir-test upgrade mimir grafana/mimir-distributed -f custom.yaml
   ```

> 1. From [Grafana Explore](http://localhost:3000/explore), verify that your metrics are being scraped, by querying `sum(rate(cortex_ingester_ingested_samples_total[$__rate_interval]))`.

Reviewer 3: Metrics can still be getting scraped, but this query can return no data or `0` if remote-write isn't set up correctly.

```suggestion
1. From [Grafana Explore](http://localhost:3000/explore), verify that your metrics are being written to Grafana Mimir, by querying `sum(rate(cortex_ingester_ingested_samples_total[$__rate_interval]))`.
```

Reviewer: added

---
title: "Getting started with Grafana Mimir using the Helm chart"
menuTitle: "Getting started using the Helm chart"
description: "Learn how to get started with Grafana Mimir using the Helm chart."
weight: 20
---

# Getting started with Grafana Mimir using the Helm chart

> The Helm chart allows you to configure, install, and upgrade Grafana Mimir within a Kubernetes cluster.

Author: This sentence serves to provide some content and context, to avoid having an empty bloated section.

We can add a homepage link to Helm to make this more valuable.

## Before you begin

The instructions that follow are common across any flavor of Kubernetes. They also assume that you know how to install a Kubernetes cluster, and configure and operate it.

> It also assumes that you have an understanding of what the `kubectl` command does.

Reviewer: not really, we show specific Helm commands and don't ask the user to be proficient with it.

> It also assumes that you have an understanding of what the `kubectl` command does.

Author: Explaining what Helm is in more detail is available via Helm documentation. We provide a link to the definitive source of information.

> **Caution:** Do not use this getting-started procedure in a production environment.

Hardware requirements:

> - A single Kubernetes node with a minimum of 4 cores and 16GiB RAM

Author: Thank you. :) If there is an action to take, we can incorporate it.

Software requirements:

> - Kubernetes 1.10 or higher

Reviewer: The Chart has 1.10. We might as well bump it to 1.20, we certainly don't test with anything that old, but it needs to be in the Changelog as well.

> - Kubernetes 1.10 or higher

Author: @krajorama is going to address this in a separate PR.

- The `kubectl` command for your version of Kubernetes
- Helm 3 or higher

Verify that you have:

- Access to the Kubernetes cluster
- Persistent storage is enabled in the Kubernetes cluster, which has a default storage class set up
- DNS service works in the Kubernetes cluster
- An ingress controller is set up in the Kubernetes cluster

>   **Note:** Although this is not strictly necessary, if you want to access Mimir from outside of the Kubernetes cluster, you will need an ingress. This procedure assumes you have an ingress controller set up.

Reviewer 5: ```suggestion
> **Note:** Although this is not strictly necessary, if you want to access Mimir from outside of the Kubernetes cluster, you will need an ingress. This procedure assumes you have an ingress controller set up.
```

1. Set up a Helm repository using the following commands:

   ```console
   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

   > **Note:** The Helm chart at [https://grafana.github.io/helm-charts](https://grafana.github.io/helm-charts) is a publication of the source code at [**grafana/mimir**](https://github.com/grafana/mimir/tree/main/operations/helm/charts/mimir-distributed).

1. Configure an ingress:

   a. Create a YAML file of Helm values called `custom.yaml`.

   b. Add the following configuration to the file:

   ```yaml
   nginx:
     ingress:
       enabled: true
       ingressClassName: nginx
       hosts:
         - host: <ingress-host>
           paths:
             - path: /
               pathType: Prefix
       tls:
         # empty, disabled.
   ```

   An ingress enables you to externally access a Kubernetes cluster.
   Replace _`<ingress-host>`_ with a suitable hostname that DNS can resolve to the external IP address of the Kubernetes cluster. For more information, see [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/).

1. Install Grafana Mimir using the Helm chart:

   ```bash

>    helm -n <namespace> install <release-name> grafana/mimir-distributed -f custom.yaml

Reviewer: We replaced the templates with hardcoded values now.

   ```

> 1. Check the statuses of the Mimir services:

Reviewer 5: Service is a different entity in K8S, so I would call it "pods" here.

```suggestion
1. Check the statuses of the Mimir pods:
```

   ```bash

>    kubectl -n <namespace> get pod

Reviewer: done

   ```

   The results look similar to this:

   ```bash
   kubectl -n dev get pod
   NAME                                                       READY   STATUS      RESTARTS   AGE

>    <release-name>-mimir-nginx-69fd969c-g6bkn                  1/1     Running     0          159m

Reviewer: should be fixed now because we switched to hardcoded release name

   <release-name>-mimir-distributor-7db5f7c8bc-444sd          1/1     Running     0          148m
   <release-name>-mimir-overrides-exporter-79b475c98d-sqzxx   1/1     Running     0          148m
   <release-name>-mimir-querier-5cb58b7b9c-zwnr6              1/1     Running     0          148m
   <release-name>-mimir-query-frontend-74d666bcd5-4dcwh       1/1     Running     0          148m
   <release-name>-mimir-ruler-745fff6c6d-8swsv                1/1     Running     0          148m
   <release-name>-mimir-alertmanager-0                        1/1     Running     0          148m
   <release-name>-mimir-compactor-0                           1/1     Running     0          148m
   <release-name>-mimir-ingester-1                            1/1     Running     0          147m
   <release-name>-mimir-ingester-2                            1/1     Running     0          147m
   <release-name>-mimir-store-gateway-0                       1/1     Running     0          146m
   <release-name>-mimir-ingester-0                            1/1     Running     0          144m
   ```

1. What until all of the pods have a status of `Running` or `Completed`, which might take a few minutes.

> ## Configure Prometheus to write to Grafana Mimir

Author: Duplicate feedback, so resolving this one first.

Make a choice based on whether or not you already have a Prometheus server set up:

- For an existing Prometheus server:

  1. Add the following YAML snippet to your Prometheus configuration file:

     ```yaml
     remote_write:
       - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Prometheus server writes metrics to Grafana Mimir, based on what is defined in the existing `scrape_configs` configuration.

  1. Restart the Prometheus server.

- For a Prometheus server that does not exist yet:

  1. Write the following configuration to a `prometheus.yml` file:

     ```yaml
     remote_write:
       - url: http://<ingress-host>/api/v1/push

     scrape_configs:
       - job_name: prometheus
         honor_labels: true
         static_configs:
           - targets: ["localhost:9090"]
     ```

     In this case, your Prometheus server writes metrics to Grafana Mimir that it scrapes from itself.

  1. Start a Prometheus server by using Docker:

     ```bash
     docker run --network=host -p 9090:9090  -v <path-to>/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
     ```

> ## Configure Grafana Agent to write to Grafana Mimir

Reviewer 5: I would clarify that you don't have to run both Prometheus and Grafana Agent, but you can run Grafana Agent instead of Prometheus.

     remote_write:
       - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Grafana Agent will write metrics to Grafana Mimir, based on what is defined in the existing `metrics.configs.scrape_configs` configuration.

  1. Restart the Grafana Agent.

- For a Grafana Agent that does not exist yet:

  1. Write the following configuration to an `agent.yaml` file:

     ```yaml
     server:
       http_listen_port: 12345
       grpc_listen_port: 54321

     metrics:
       wal_directory: /tmp/grafana-agent/wal

       configs:
         - name: agent
           scrape_configs:
             - job_name: agent
               static_configs:
                 - targets: ["127.0.0.1:12345"]
           remote_write:
             - url: http://<ingress-host>/api/v1/push
     ```

     In this case, your Grafana Agent writes metrics to Grafana Mimir that it scrapes from itself.

  1. Create an empty directory for the write ahead log (WAL) of the Grafana Agent

  1. Start a Grafana Agent by using Docker:

     ```bash
     docker run --network=host  -v <path-to-wal-directory>:/etc/agent/data -v <path-to>/agent.yaml:/etc/agent/agent.yaml grafana/agent
     ```

> ## Query data in Grafana

Reviewer 5: Should we just be more specific and call it "metrics"?

```suggestion
## Query metrics in Grafana
```

First install Grafana, and then add Mimir as a Prometheus data source.

1. Start Grafana by using Docker:

   ```bash
   docker run --rm --name=grafana --network=host grafana/grafana
   ```

1. In a browser, go to the Grafana server at [http://localhost:3000](http://localhost:3000).
1. Sign in using the default username `admin` and password `admin`.
1. On the left-hand side, go to **Configuration** > **Data sources**.

> 1. Configure a new Prometheus data source to query the local Grafana Mimir server, by using the following settings:

Reviewer 5: ```suggestion
1. Configure a new Prometheus data source to query the local Grafana Mimir cluster, by using the following settings:
```

   | Field | Value                              |
   | ----- | ---------------------------------- |
   | Name  | Mimir                              |
   | URL   | http://\<ingress-host\>/prometheus |

   To add a data source, see [Add a data source](https://grafana.com/docs/grafana/latest/datasources/add-a-data-source/).

1. Verify success:

>    You should be able to query metrics in [Grafana Explore](https://grafana.com/docs/grafana/latest/explore/),

Reviewer 5: I would consider linking the Grafana running locally.

```suggestion
   You should be able to query metrics in [Grafana Explore](http://localhost:3000/explore),
```

>    as well as create dashboard panels by using your newly configured `Mimir` data source.

Reviewer 5: I would consider linking the "Monitoring Grafana Mimir" documentation page to install Mimir dashboards.

Reviewer: done

## Set up metamonitoring

Grafana Mimir metamonitoring collects metrics or logs, or both,
about Grafana Mimir itself.
In the example that follows, metamonitoring scrapes metrics about
Grafana Mimir itself, and then writes those metrics to the same Grafana Mimir instance.

1. To enable metamonitoring in Grafana Mimir, add the following YAML snippet to your Grafana Mimir `custom.yaml` file:

   ```yaml
   metaMonitoring:
     serviceMonitor:
       enabled: true
     grafanaAgent:
       enabled: true
       installOperator: true
       metrics:
         additionalRemoteWriteConfigs:

>            - url: "http://<release-name>-mimir-nginx.<namespace>.svc:80/api/v1/push"

Reviewer: Yes, this turned out to be a weird quirk inherited from the original GEM chart, fixed by specifying the name. Plus the chart will print the correct URLs on install/upgrade now. Ref: https://github.com/grafana/mimir/issues/2282

   ```

   Replace _`<release-name>`_ with the release name that you used when you installed the Grafana Mimir Helm chart.

   Replace _`<namespace>`_ with the namespace in which Grafana Mimir is installed.

1. Upgrade Grafana Mimir by using the `helm` command:

   ```bash
   helm -n <namespace> upgrade <release-name> grafana/mimir-distributed -f custom.yaml
   ```

> 1. In Grafana, verify that your metrics are being scraped, by quering the metric `cortex_ingester_ingested_samples_total{}`.

Reviewer 5: I would provide a full query to just copy-paste:

```suggestion
1. In Grafana, verify that your metrics are being scraped, by quering `sum(rate(cortex_ingester_ingested_samples_total[$__rate_interval]))`.
```

> 

Reviewer: We could include something for k3d, it's very popular and in the chart README (which we eventually want to get rid of) we have some help regarding that: https://github.com/grafana/mimir/blob/main/operations/helm/charts/mimir-distributed/README.md#development

> ## Query data in Grafana that is running within the same Kubernetes cluster

Reviewer 5: ```suggestion
## Query metrics in Grafana that is running within the same Kubernetes cluster
```

1. Install Grafana in the same Kubernetes cluster.

   For details, see [Deploy Grafana on Kubernetes](https://grafana.com/docs/grafana/latest/setup-grafana/installation/kubernetes/).

1. Stop the Grafana instance that is running in the Docker container, to allow for port-forwarding.

1. Port-forward Grafana to `localhost`, by using the `kubectl` command:

   ```bash
   kubectl port-forward service/grafana 3000:3000
   ```

1. In a browser, go to the Grafana server at [http://localhost:3000](http://localhost:3000).
1. Sign in using the default username `admin` and password `admin`.
1. On the left-hand side, go to **Configuration** > **Data sources**.
1. Configure a new Prometheus data source to query the local Grafana Mimir server, by using the following settings:

   | Field | Value                                                               |
   | ----- | ------------------------------------------------------------------- |
   | Name  | Mimir                                                               |

>    | URL   | http://\<release-name\>-mimir-nginx.\<namespace\>.svc:80/prometheus |

Reviewer: Yes, this turned out to be a weird quirk inherited from the original GEM chart, fixed by specifying the name. Plus the chart will print the correct URLs on install/upgrade now. Ref: https://github.com/grafana/mimir/issues/2282

1. Install Grafana Mimir using the Helm chart:

   ```bash
   helm -n mimir-test install mimir grafana/mimir-distributed -f custom.yaml
   ```

   > **Note:**: the output of the command contains the write and read URLs necessary for the following steps.

1. Check the statuses of the Mimir pods:

   ```bash
   kubectl -n mimir-test get pod
   ```

   The results look similar to this:

   ```bash
   kubectl  -n mimir-test get pod
   NAME                                            READY   STATUS      RESTARTS   AGE
   mimir-minio-78b59f5569-fhlhs                    1/1     Running     0          2m4s
   mimir-nginx-74f8bff8dc-7kr7z                    1/1     Running     0          2m5s

>    mimir-mimir-distributed-make-bucket-job-z2hc8   0/1     Completed   0          2m4s

Reviewer 6: Looks like you missed this one (extra mimir-)

Author: ```suggestion
   mimir-distributed-make-bucket-job-z2hc8   0/1     Completed   0          2m4s
```

Reviewer: it's a tiny bug : https://github.com/grafana/helm-charts/issues/1345

Reviewer: but we'll fix it in the doc anyway
